File: backend/app/services/ingredient_quantity.py
"""Разбор количества в строках ингредиентов и объединение для списка покупок."""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Dict, List, Optional, Tuple

# (regex, unit_key, display_unit)
_QUANTITY_PATTERNS: List[Tuple[str, str, str]] = [
    (r"(\d+(?:[.,]\d+)?)\s*(?:кг|kg)\b", "kg", "кг"),
    (r"(\d+(?:[.,]\d+)?)\s*(?:г|гр|грамм\w*)\b", "g", "г"),
    (r"(\d+(?:[.,]\d+)?)\s*(?:мл|ml)\b", "ml", "мл"),
    (r"(\d+(?:[.,]\d+)?)\s*(?:л|l)\b", "l", "л"),
    (r"(\d+)\s*(?:шт\.?|штук\w*)\b", "pcs", "шт"),
    (r"(\d+)\s*(?:ст\.?\s*л\.?|столов\w*\s*ложк\w*)\b", "tbsp", "ст. л."),
    (r"(\d+)\s*(?:ч\.?\s*л\.?|чайн\w*\s*ложк\w*)\b", "tsp", "ч. л."),
    (r"(\d+)\s*(?:зубч\.?|зубчик\w*)\b", "clove", "зуб."),
    (r"(\d+)\s*(?:пучок|пучк\w*)\b", "bunch", "пучок"),
    (r"(\d+)\s*(?:стак\.?|стакан\w*)\b", "cup", "стак."),
]


@dataclass
class ParsedIngredient:
    name: str
    amount: Optional[float] = None
    unit: Optional[str] = None  # g, ml, pcs, ...
    portions: int = 1

    @property
    def merge_key(self) -> str:
        base = _normalize_name(self.name)
        if self.unit:
            return f"{base}|{self.unit}"
        return base

    def scaled(self, factor: float) -> "ParsedIngredient":
        if self.amount is None or factor == 1:
            return self
        return ParsedIngredient(
            name=self.name,
            amount=round(self.amount * factor, 1),
            unit=self.unit,
        )

    def display_quantity(self) -> Optional[str]:
        if self.amount is None or not self.unit:
            if self.portions > 1:
                return f"×{self.portions} порц."
            return None
        val = self.amount
        if val == int(val):
            val_s = str(int(val))
        else:
            val_s = f"{val:g}".replace(".", ",")
        unit_labels = {
            "g": "г",
            "kg": "кг",
            "ml": "мл",
            "l": "л",
            "pcs": "шт",
            "tbsp": "ст. л.",
            "tsp": "ч. л.",
            "clove": "зуб.",
            "bunch": "пучок",
            "cup": "стак.",
        }
        return f"{val_s} {unit_labels.get(self.unit, self.unit)}"


def _normalize_name(name: str) -> str:
    s = name.strip().lower()
    s = re.sub(r"\s+", " ", s)
    return s


def parse_ingredient_line(line: str) -> ParsedIngredient:
    """«куриная грудка 200 г» → name + amount."""
    raw = line.strip()
    if not raw:
        return ParsedIngredient(name="")

    working = raw
    amount: Optional[float] = None
    unit: Optional[str] = None

    for pattern, unit_key, _disp in _QUANTITY_PATTERNS:
        m = re.search(pattern, working, flags=re.IGNORECASE)
        if not m:
            continue
        amount = float(m.group(1).replace(",", "."))
        unit = unit_key
        if unit_key == "kg":
            amount *= 1000
            unit = "g"
        if unit_key == "l":
            amount *= 1000
            unit = "ml"
        working = (working[: m.start()] + working[m.end() :]).strip(" ,-–—")
        break

    name = re.sub(r"\s+", " ", working).strip(" ,-–—")
    if not name:
        name = raw
    return ParsedIngredient(name=name, amount=amount, unit=unit)


def merge_ingredient_lines(
    lines: List[str],
    *,
    portions: int = 1,
) -> Dict[str, ParsedIngredient]:
    """Объединить ингредиенты плана: суммировать граммы/штуки по продукту."""
    merged: Dict[str, ParsedIngredient] = {}
    factor = max(1, portions)

    for line in lines:
        parsed = parse_ingredient_line(line)
        if parsed.amount is not None and parsed.unit:
            parsed = parsed.scaled(factor)
        else:
            parsed = ParsedIngredient(
                name=parsed.name,
                amount=None,
                unit=None,
                portions=factor,
            )
        key = parsed.merge_key
        if not parsed.name:
            continue
        if key not in merged:
            merged[key] = parsed
            continue
        prev = merged[key]
        if (
            prev.amount is not None
            and parsed.amount is not None
            and prev.unit == parsed.unit
        ):
            merged[key] = ParsedIngredient(
                name=prev.name,
                amount=prev.amount + parsed.amount,
                unit=prev.unit,
                portions=prev.portions,
            )
        elif parsed.amount is None and prev.amount is None:
            merged[key] = ParsedIngredient(
                name=prev.name,
                amount=None,
                unit=None,
                portions=prev.portions + parsed.portions,
            )
        else:
            # разные единицы — оставляем отдельной строкой с суффиксом
            alt_key = f"{key}#{len(merged)}"
            merged[alt_key] = parsed

    return merged

File: lib/utils/ingredient_quantity.dart
/// Разбор количества в строках ингредиентов (г, кг, мл, шт).
class ParsedIngredient {
  const ParsedIngredient({
    required this.name,
    this.amount,
    this.unit,
    this.portions = 1,
  });

  final String name;
  final double? amount;
  final String? unit;
  final int portions;

  String? get displayQuantity {
    if (amount != null && unit != null) {
      final val = amount!;
      final valS = val == val.roundToDouble()
          ? '${val.round()}'
          : val.toStringAsFixed(1).replaceAll('.', ',');
      return '$valS ${_unitLabel(unit!)}';
    }
    if (portions > 1) return '×$portions порц.';
    return null;
  }

  static String _unitLabel(String unit) => switch (unit) {
        'g' => 'г',
        'ml' => 'мл',
        'pcs' => 'шт',
        'tbsp' => 'ст. л.',
        'tsp' => 'ч. л.',
        'clove' => 'зуб.',
        'bunch' => 'пучок',
        'cup' => 'стак.',
        _ => unit,
      };
}

final _patterns = <({RegExp re, String unit, bool scaleToBase})>[
  (re: RegExp(r'(\d+(?:[.,]\d+)?)\s*(?:кг|kg)\b', caseSensitive: false), unit: 'g', scaleToBase: true),
  (re: RegExp(r'(\d+(?:[.,]\d+)?)\s*(?:г|гр|грамм\w*)\b', caseSensitive: false), unit: 'g', scaleToBase: false),
  (re: RegExp(r'(\d+(?:[.,]\d+)?)\s*(?:мл|ml)\b', caseSensitive: false), unit: 'ml', scaleToBase: false),
  (re: RegExp(r'(\d+(?:[.,]\d+)?)\s*(?:л|l)\b', caseSensitive: false), unit: 'ml', scaleToBase: true),
  (re: RegExp(r'(\d+)\s*(?:шт\.?|штук\w*)\b', caseSensitive: false), unit: 'pcs', scaleToBase: false),
  (re: RegExp(r'(\d+)\s*(?:ст\.?\s*л\.?|столов\w*\s*ложк\w*)\b', caseSensitive: false), unit: 'tbsp', scaleToBase: false),
  (re: RegExp(r'(\d+)\s*(?:ч\.?\s*л\.?|чайн\w*\s*ложк\w*)\b', caseSensitive: false), unit: 'tsp', scaleToBase: false),
  (re: RegExp(r'(\d+)\s*(?:зубч\.?|зубчик\w*)\b', caseSensitive: false), unit: 'clove', scaleToBase: false),
];

ParsedIngredient parseIngredientLine(String line) {
  var working = line.trim();
  if (working.isEmpty) return const ParsedIngredient(name: '');

  double? amount;
  String? unit;

  for (final p in _patterns) {
    final m = p.re.firstMatch(working);
    if (m == null) continue;
    amount = double.parse(m.group(1)!.replaceAll(',', '.'));
    if (p.scaleToBase) {
      amount = p.unit == 'g' ? amount * 1000 : amount * 1000;
    }
    unit = p.unit;
    working = '${working.substring(0, m.start)}${working.substring(m.end)}'
        .trim()
        .replaceAll(RegExp(r'^[\s,\-–—]+'), '')
        .replaceAll(RegExp(r'[\s,\-–—]+$'), '');
    break;
  }

  var name = working.replaceAll(RegExp(r'\s+'), ' ').trim();
  if (name.isEmpty) name = line.trim();
  return ParsedIngredient(name: name, amount: amount, unit: unit);
}

String _normalizeName(String name) =>
    name.trim().toLowerCase().replaceAll(RegExp(r'\s+'), ' ');

/// Объединить строки ингредиентов для списка покупок.
List<({String name, String? quantity})> mergeIngredientLines(
  List<String> lines, {
  int portions = 1,
}) {
  final factor = portions < 1 ? 1 : portions;
  final merged = <String, ParsedIngredient>{};

  for (final line in lines) {
    var parsed = parseIngredientLine(line);
    if (parsed.amount != null && parsed.unit != null) {
      if (factor != 1) {
        parsed = ParsedIngredient(
          name: parsed.name,
          amount: (parsed.amount! * factor * 10).round() / 10,
          unit: parsed.unit,
        );
      }
    } else {
      parsed = ParsedIngredient(
        name: parsed.name,
        portions: factor,
      );
    }
    final key = parsed.unit != null
        ? '${_normalizeName(parsed.name)}|${parsed.unit}'
        : _normalizeName(parsed.name);
    if (parsed.name.isEmpty) continue;

    final prev = merged[key];
    if (prev == null) {
      merged[key] = parsed;
      continue;
    }
    if (prev.amount != null &&
        parsed.amount != null &&
        prev.unit == parsed.unit) {
      merged[key] = ParsedIngredient(
        name: prev.name,
        amount: prev.amount! + parsed.amount!,
        unit: prev.unit,
      );
    } else if (prev.amount == null && parsed.amount == null) {
      merged[key] = ParsedIngredient(
        name: prev.name,
        portions: prev.portions + parsed.portions,
      );
    } else {
      merged['$key#${merged.length}'] = parsed;
    }
  }

  return merged.values
      .map((p) => (
            name: p.name.isEmpty
                ? p.name
                : p.name[0].toUpperCase() + p.name.substring(1),
            quantity: p.displayQuantity,
          ))
      .toList();
}
